File: Semestrul 5/PPL/Laborator1/src/main/java/ro/ppd2024/distributor/CyclicDistributor.java
package ro.ppd2024.distributor;


import ro.ppd2024.thread.cyclic.DeltaCyclicConvolutionThread;
import ro.ppd2024.thread.cyclic.HorizontalCyclicConvolutionThread;
import ro.ppd2024.thread.cyclic.VerticalCyclicConvolutionThread;
import ro.ppd2024.util.Technique;

public class CyclicDistributor implements Distributor {
    @Override
    public void distribute(int[][] F, int n, int m, int[][] V, int[][] C, int k,
                                  int nrThreads, Technique technique) {
        Thread[] threads = new Thread[nrThreads];

        for (int i = 0; i < nrThreads; i++) {
            switch (technique) {
                case HORIZONTAL_CYCLIC:
                    threads[i] = new HorizontalCyclicConvolutionThread(F, V, n, m, C, k, i, nrThreads);
                    break;
                case VERTICAL_CYCLIC:
                    threads[i] = new VerticalCyclicConvolutionThread(F, V, n, m, C, k, i, nrThreads);
                    break;
                case DELTA_CYCLIC:
                    threads[i] = new DeltaCyclicConvolutionThread(F, V, n, m, C, k, i, nrThreads);
                    break;
                default:
                    throw new IllegalArgumentException("Invalid technique");
            }
            threads[i].start();
        }

        for (int i = 0; i < nrThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Semestrul 5/PPD/Laborator1/ConvolutionCpp/distributor/CyclicDistributor.cpp
#include "CyclicDistributor.h"
#include "../thread/CyclicConvolutionThread.h"
#include <thread>
#include <stdexcept>

template <typename TMatrix, typename TFilter>
void CyclicDistributor<TMatrix, TFilter>::distribute(TMatrix F, int n, int m, TMatrix V, TFilter C, int k, int nrThreads, Technique technique) {
    std::thread* threads = new std::thread[nrThreads];

    for (int i = 0; i < nrThreads; i++) {
        switch (technique) {
            case Technique::HORIZONTAL_CYCLIC:
                threads[i] = std::thread(horizontalCyclicConvolutionThread, F, V, n, m, C, k, i, std::min(nrThreads, n));
                break;
            case Technique::VERTICAL_CYCLIC:
                threads[i] = std::thread(verticalCyclicConvolutionThread, F, V, n, m, C, k, i, std::min(nrThreads, m));
                break;
            case Technique::DELTA_CYCLIC:
                threads[i] = std::thread(deltaCyclicConvolutionThread, F, V, n, m, C, k, i, nrThreads);
                break;
            default:
                throw std::invalid_argument("Invalid technique");
        }
    }

    for (int i = 0; i < nrThreads; i++) {
        threads[i].join();
    }
    delete[] threads;
}
